src/librustc/ty/layout.rs

```diff
@@ -1710,15 +1710,74 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                     value: Int(ity, signed),
                     valid_range: (min as u128 & tag_mask)..=(max as u128 & tag_mask),
                 };
-                let abi = if tag.value.size(dl) == size {
-                    Abi::Scalar(tag.clone())
-                } else {
-                    Abi::Aggregate { sized: true }
-                };
+                let mut abi = Abi::Aggregate { sized: true };
+                if tag.value.size(dl) == size {
+                    abi = Abi::Scalar(tag.clone());
+                } else if size > dl.pointer_size {
```

Comment:
This should *really* have a `// HACK(nox)` comment on it explaining what optimizations removing the condition breaks, etc., and/or pointing to an issue.

Reply:
Done!

Reply:
Hmm, is this problem specific to bool tags? You could check `tag.is_bool()` directly, such that e.g. a three-variant that has a byte-sized tag would still get a pair.

Reply:
I went for the most conservative trade-off to make some progress, but I'll change it accordingly to your suggestion.